PinetimeFlasher.pyw

```diff
@@ -235,9 +250,82 @@ class ConfDialog(QDialog):
         except OSError:
             self.status.setText("Unable to write configuration file!")
 
-    def infoButton(self, s):
-        dlg = InfoDialog()
-        dlg.exec()
+    def setup_openocd(self):
+        """
+        Download and unpack OpenOCD for the current platform, then add it to the system path.
+        """
+        self.status.setText("Finding latest OpenOCD release...")
+
+        base_url = "https://api.github.com/repos"
+        response = requests.get(f"{base_url}/xpack-dev-tools/openocd-xpack/releases/latest")
+
+        details = response.content.decode('utf-8')
+        content = json.loads(details)
+
+        archive_file, hash_file = self.get_github_assets(content["assets"])
+        computed_hash, provided_hash = self.get_hashes(archive_file, hash_file)
+        if computed_hash != provided_hash:
+            self.status.setText("Hashes do not match - corrupted download!")
+            return
+
+        self.unpack_archive(archive_file)
+        add_openocd_to_system_path(self.status)
+        os.remove(archive_file)
+        os.remove(hash_file)
+
+    def get_hashes(self, archive_file, hash_file):
+        self.status.setText("Computing hashes OpenOCD...")
+        with open(archive_file, "rb") as fd:
+            hasher = hashlib.sha256()
+            hasher.update(fd.read())
+            computed_hash = hasher.hexdigest()
+        with open(hash_file, "r") as fd:
+            provided_hash= fd.read().split(" ")[0]   # format: "<hash> <filename>"
+        return computed_hash, provided_hash
+
+    def unpack_archive(self, archive):
+        """
+        Unpack the archive and shift files so that the files can always be found
+        under 'openocd/' (i.e. without a version number, which is how they are
+        currently packed).
+        """
+        self.status.setText("Unpacking OpenOCD...")
+        tmpdir_name = "openocd_tmp"
+        shutil.unpack_archive(archive, extract_dir=tmpdir_name)
+        tmpdir_contents = os.listdir(tmpdir_name)
+
+        if len(tmpdir_contents) == 1:
+            shutil.move(os.path.join(tmpdir_name, tmpdir_contents[0]), "openocd")
+            os.rmdir(tmpdir_name)
+        else:
+            shutil.move(tmpdir_name, "openocd")
+
+    def get_github_assets(self, assets: List[str]):
+        plat = {
+            "Windows": "win32",
+            "Linux": "linux",
+            "MacOS": "darwin",
+        }.get(platform.system(), "")
+        arch = {
+            "x86_64": "x64",
+            "i386": "ia32"
+        }.get(platform.machine(), "")
+
+        if not plat or not arch:
```

Comment:
Gotcha (I think)... `if not plat or not arch`, when fired, isn't not returning a list... so on returning it breaks.

Reply:
And this is because

plat =  Windows
arch =  AMD64

Reply:
Changing ln#310 from `x86_64` to `AMD64` cured the problem... but still means there's a failure case that isn't catered for. Once that was changed, everything worked perfect (using the portable openocd rather than a version in the system path).